Port the balance-text plugin to be VanillaJS

BrianGenisio opened the pull request:
The jQuery plugin still exists, but it is only a shell around the native version.

I have tested this against the following browsers:

- Sarari 4
- Chrome 16
- Firefox 10
- IE 9
- Edge 13

redmunds commented:
@BrianGenisio I haven't had a chance to take a close look and try this out, but this looks **awesome**! I'm pretty swamped with my day job, but I should be able to get to this over winter break (if not before). Thanks.

BrianGenisio commented:
@redmunds Awesome :)  Let me know if there is anything I can help with.  Ideally, this will get published via NPM.  I'm taking my fork as an NPM dependency right now for a React component I'm wrapping it with.  But I'm working on this for my day job (@rileyjshaw is my co-worker), so if there is anything I can do to help get this landed, I'm happy to do it :)

In the mean time, of course, no hurry.  I'm working fine with our fork for the time being!

redmunds commented:
@BrianGenisio This looks good. I thought I found a bug where first (text-justify) paragraph in sample file wasn't balancing on initial load in Safari/Chrome, but I notice it also happens in master :) Oh, well. I do have a few minor comments coming about source formatting, etc.

Also, version number needs to be incremented. Since this is a major change, I was wondering if I should it should go straight to v3.0.0 even though no API were broken?
cc: @rileyjshaw

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,513 @@
+    module.exports = factory();
+  } else {
+    root.balanceText = factory();
+  }
```
Current code satisfies JSLint, so please run that. Most errors are source formatting.

redmunds replied:
This directive eliminates some errors:

```
/*global define, module, CustomEvent, NodeList */
```

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,513 @@
+            document.addEventListener('DOMContentLoaded', fn);
+        } else {
+            document.attachEvent('onreadystatechange', function() {
+            if (document.readyState != 'loading')
```
Missing `{` at end of this line

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,513 @@
+            document.attachEvent('onreadystatechange', function() {
+            if (document.readyState != 'loading')
+                fn();
+            });
```
Errant `);` ?

redmunds replied:
Oh, I see it's for `document.attachEvent();`

Need to indent `if` statement in function.

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,513 @@
+        }
+    }
+
+    function trigger(el, event, data) {
```
Param `event` is re-declared as a var.

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,513 @@
+
+    var removeTags = function (el) {
+        var brs = nodeListAsArray(el.querySelectorAll('br[data-owner="balance-text"]'));
+        brs.forEach(function(br) { return br.outerHTML = " "; });
```
Is `return` necessary?

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,513 @@
+                span.parentNode.removeChild(span)
+            });
+            el.innerHTML = txt;
+            console.log(txt);
```
Remove `console.log()`

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,513 @@
+
+        elements = ensureElementArray(elements);
+
+        if (hasTextWrap) {
```
This check for quick exit should be first thing done in function

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,513 @@
+            var txt = "";
+            spans.forEach(function(span) {
+                txt += span.textContent;
+                span.parentNode.removeChild(span)
```
Missing semicolon

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,513 @@
+
+        container.appendChild(space);
+
+        el.appendChild(container)
```
Missing semicolon

BrianGenisio commented:
@redmunds Thanks for the review!  Yes, I saw that there were some issues, but I chose to avoid fixing them in this PR in favor of it not introducing any behavioral changes whatsoever.  The fact that the bugs persist was a validation that the port works :)

I do think that it should bump to 3.0.0 (I've updated package.json to reflect that).  There IS a breaking change: you now need to include both files in your distribution (the native lib and the jQuery wrapper).  Seems like a major version bump is prudent.

Regarding that, I don't see this package published in npm.  Do you want to do that publish?  If not, I'm happy to do it (and add you as a collaborator).  I have a (`react-balance-text`)[https://github.com/khan/react-balance-text] package waiting to be published, which will require this package.

I've updated this PR with your suggestions.  It runs clean with `jslint` (I missed the jslint comments in the header.  Thanks for pointing that out) and I ran `jsmin` on the files.

After this, I recommend updating it a bit with some `npm run` scripts for running `jslint` and minification, (es2015?), etc.  If you are open to it, I'm happy to send that PR across :)

redmunds commented:
@BrianGenisio

> I don't see this package published in npm.

@bfred-it has published it at: https://www.npmjs.com/package/balance-text

See this issue: https://github.com/adobe-webplatform/balance-text/issues/84#issuecomment-233617975

BrianGenisio commented:
Ah, ok.  I see it now.  The NPM auto-complete didn't show it (which is odd).  That's great.  So yeah, once this is landed, we should get it published to NPM so that I can lean on the dependency.

Thanks!
B

bfred-it commented:
I'll be waiting to publish it 👍

bfred-it commented:
One thing though. Having to load two separate files is sure to cause issues for people. We should put a bundler in place. I can make a PR to bundle it with rollup and setup the npm package to be easily usable. Actually, it might be an even better idea to have `jquery-balance-text` as its own module. 

If agreed I'll push a PR for this change after this one is in.

BrianGenisio commented:
@bfred-it My two cents: breaking out `jquery-balance-text` makes a ton of sense.  There is already `angular-balance-text` and `react-balance-text` is on the heels of this landing, so `jquery-balance-text` would fit that pattern nicely.  

The only problem is logistical.  People expecting the jQuery integration would be looking for it here. By taking out the jQuery functionality from this particular package, it could cause confusion... but perhaps that is OK with a major version bump and good documentation?

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,525 @@
+ * - Chrome 16
+ * - Firefox 10
+ * - IE 9
+ * - Edge 13
```
Should remove this list of browser versions tested against since this will most likely not be kept up-to-date.

redmunds commented on balancetext.js:
```diff
@@ -0,0 +1,525 @@
+     * the specified size.
+     *
+     * @param txt      - text string
+     * @param conWidth - container width
```
Missing first param:

```
     * @param el        - element
```

bfred-it commented:
The repo wouldn't change, but the npm package would.

 With your PR, the users of `balance-text` will already have to use a weird require like `balance-text/jquery.balance-text` for the jquery version, so we might as well just extract it into its own module and make it simpler.

BrianGenisio commented:
@redmunds I just updated all of the function comments/params (several were out of date as well).  I also added two NPM scripts: `npm run lint` and `npm run build` (only minifies now).  I kept the references to the lowest browsers, since I think they are valuable for when people make changes in the future.  Again, I'll defer to you if you still think they should be removed.

@bfred-it That seems reasonable.  Would you like to do that after this lands, but before we publish to NPM since I think you understand what you are thinking of better than I?  You can add on to the `npm run build` script.

redmunds commented on package.json:
```diff
@@ -1,37 +1,43 @@
 {
-    "name": "balance-text",
```
Looks like indenting was changed from 4 spaces to 2 spaces.

redmunds commented on package.json:
```diff
@@ -1,37 +1,43 @@
-        ]
+  "name": "balance-text",
+  "version": "3.0.0",
+  "filename": "jquery.balancetext.min.js",
```
I have this lib hosted on a couple CDNs which auto-update when a new github Release is created. The `filename` field needs to be updated for https://cdnjs.cloudflare.com and `main` needs to be updated for https://cdn.jsdelivr.net to `balancetext.min.js`.

The thing I'm not quite sure about is whether exsiting libs for version <= 2.0.0 will be maintained. Does anyone know how the CDN auto-updating  work?

bfred-it commented:
I gave a proper look to [jquery.balancetext.js](https://github.com/adobe-webplatform/balance-text/blob/6ce6f310459da4430f2a7083056385c77a5e3632/jquery.balancetext.js) and we can/probably-should just inline that in balancetext.js instead of having it as a separate file/module/package.

@BrianGenisio can you add it [here](https://github.com/adobe-webplatform/balance-text/blob/6ce6f310459da4430f2a7083056385c77a5e3632/balancetext.js#L585)? Like `if(window.jQuery){ /* attach jQuery.fn........ */}`

BrianGenisio commented:
@bfred-it I'm moving that code over and I've come to a sticky bit at [this line](https://github.com/adobe-webplatform/balance-text/blob/6ce6f310459da4430f2a7083056385c77a5e3632/jquery.balancetext.js#L72).  

At that line, assuming the jQuery functionality, we add the watchers for `document.ready` and `window.resize`.  That is how the inclusion of the jQuery function worked, but it is not how I had envisioned the VanillaJS module working.  I'd like that module to support either case.

By automatically hooking up the jQuery portion of it, we are forcing them into a usage pattern regardless of how they want to use the library.  For example, I'd like the react-balance-text component to only do this at the component level and not hook the global handlers.  But jQuery IS available in my app, (even though we are phasing it out) and this would introduce unintended side affects.

We _could_ change the jQuery watching to be opt-in (which would be breaking change in 3.0.0), requiring jQuery users to do something like `$.balanceText()` or some other call to denote that we want to balance the page and watch for window resizes.

BrianGenisio commented:
(This reminds me that I need to update the README.md to describe the library in general terms.  I'm on that as well.)

BrianGenisio commented:
@bfred-it and @redmunds 

I just sent through another set of changes.  I've merged the two files into one and (for the time being, at least) now it requires an explicit opt-in `$.balanceText()` for jQuery users.  This is a breaking change, but I don't know any other way to bring both libraries together.  I also updated the `README.md` to explain the Vanilla option, and other inconsistencies now that jQuery is an option, not a requirement.

I'm open to other ideas if anyone has any.

redmunds commented:
@BrianGenisio Seems like it would be better to keep jQuery plugin code in a separate file (since that code will rarely change) than to change the API, unless @bfred-it knows of a better way to combine the files.

bfred-it commented:
@BrianGenisio So the jQuery version is automatically enabled for `.balance-text` elements but the vanilla version isn't? I wouldn't make that distinction. Either BT is an automatic polyfill or it isn't. jQuery shouldn't make that difference. 

@redmunds Keeping it as a separate file complicates things for everyone though and I don't see any advantages since it's like 6 lines of code. While it may not change you'll still have to build it every time. I've seen other plugins automatically hook to jQuery if found. 

If anything, keeping jQuery in the main file means that this isn't a breaking change at all.

BrianGenisio commented:
Trying to reconcile.  Is this a polyfill?  Or is it a utility?  I suppose it is both.  In the case of "both", I see the argument for making the regular one listen to events.  I had originally thought the VanillaJS port would just be a utility and the integrations would decide whether they were polyfills or not.

For example, it seems like an anti-pattern to me to create a React component which has the side affect of affecting the entire document.  It does't feel that way for jQuery.

If we kept two files, we could just update the `npm run build` task to change from just minifying to webpack/minify.  That way the user would only include one file: the VanillaJS file OR the jQuery file (which would package with the VanillaJS version).  The `npm run build` still needs to happen every time you make a change because it still needs to minify, so that wouldn't be a big change.

I suppose I'll defer to @redmunds to see if you prefer the VanillaJS file to be a Utility or a Polyfill/Utility?  That would dictate whether we have one file which always polyfills and adds jQuery if it exists vs a VanillaJS utility and a jQuery package which includes VanillaJS and polyfills.

What is the status of the standards change?  Is it likely to get added?  If so, a polyfill seems prudent.  If not, a utility seems prudent.  (that last part are my 2c, or course)

Rycochet commented:
You have to be careful where you get jQuery from - it's not necessarily global due to RequireJS, could be Zepto, or any number of other things.

Personally I'd do something like `balanceText.polyfill(jQuery.fn)`, where the variable to be added to would be an object that a `.balanceText()` method added to it, where `this` would be an Array-like list of HTMLElement that would get balanced.

Doing it this way could let it be added to a RequireJS sourced jQuery, Zepto, or added to custom code that needed it. The amount of code needed for that would be tiny and be left as part of the main code.

Just as a FYI, this is something planned for VelocityJS v2 - not least to give an (unsafe) usage example so people can do things like `document.querySelectorAll("div").balanceText()` using `balanceText.polyfill(nodeList)`.

bfred-it commented:
@Rycochet actually I think that if we support RequireJS, we can't merge it into a single file because [`define(['jquery'])`](https://github.com/umdjs/umd/blob/95563fd6b46f06bda0af143ff67292e7f6ede6b7/templates/jqueryPlugin.js#L6) would fail unless jQuery is available. Correct me if I'm wrong.

@BrianGenisio my suggestion would be to skip the automated polyfill for all, like other polyfills do it. Let the user call `balanceText.polyfill()` (or `$.balanceText.polyfill()`)

Rycochet commented:
@bfred-it It would be up to the user to do something like -

```javascript
define(['jquery'], function(jQuery) {
   balanceText.polyfill(jQuery.fn);
});
```

As you can't really make any presumptions at all with RequireJS (not even the name of jQuery), it's sort of got to be their choice. Having it add them to a global jQuery/Zepto if they exist could still be part of it to make things "easier" for people who want to use it in the simplest sense - for the sake of just a few lines of code in there.

bfred-it commented:
@Rycochet Sure, I see, as long as don't use `define` anywhere in `balance-text`.

Also your usage of "`.polyfill`" is confusing in a CSS context. That would be `balanceText.use(jQuery)` or `balanceText.extend(nodeList)` while I'd expect `balanceText.polyfill(/* no args */)` to apply the fix globally.

BrianGenisio commented:
@bfred-it I like that approach (it is essentially the approach I've taken in the current code... I like `.pollyfill()` better than `.initHandlers()`), though it would be a breaking change from 2.0.0 since it does the polyfill implicitly.  I'll wait for @redmunds to weigh in.

And if we are moving to requiring an initialization, it would be reasonable to add @rycochet's suggestion.  If jQuery is passed in to `.pollyfill()`, then we'd hook up the jQuery wrapper.  

Honestly, I think the jQuery wrapper is only for backwards compatibility.  It doesn't provide any extra functionality anymore... so going this way is a nod towards that.  We are essentially communicating to the user that the jQuery integration has become an opt-in mechanism.

I also like that opt-in mechanism because the way it is now, it adds it as a jQuery plugin whether you want it or not (assuming it finds jQuery).  I'll be honest, I don't want it in our project's jQuery space as we are doing our best to deprecate all usage of jQuery in our project and adding the wrapper just pollutes the jQ plugin space.

BrianGenisio commented:
Ah.  I like `balanceText.use(jQuery)` and `balanceText.polyfill()` as being distinct.  Would `balanceText.use(jQuery)` also enable the polyfill?  My inclination is "yes" because it did so in 2.0.0.

Rycochet commented:
I like `balanceText.addTo(jQuery.fn)` more than `polyfill` - makes it a bit more obvious what's happening (and technically wouldn't be a polyfill) - so the name was just useful for the idea. `.use()` implies that this is using jQuery, rather than the other way around - but that's just semantics.

Having the code include a small check-and-patch for jQuery and Zepto would probably catch the majority of uses:

```javascript
if (window.jQuery && window.jQuery.fn) {
   balanceText.addTo(window.jQuery.fn);
}
if (window.Zepto&& window.Zepto.fn) {
   balanceText.addTo(window.Zepto.fn);
}
```

BrianGenisio commented:
@rycochet The jQuery portion of this is purely optional and not intended to be necessary whatsoever.  So `balanceText.polyfill()` would be the preferred mechanism to polyfill with.  `balanceText.addTo(jQuery.fn)` would be the opt-in jQuery extension, but it would only really exist for backwards compatibility.  The recommended approach, going forward, would be to not use the jQuery integration.  In the current incatation (2.0.0), it exists solely as a jQuery plugin and it acts as a polyfill.  So the question I have: should `balanceText.addTo(jQuery.fn)` also enable (implicit) the polyfill to make it compatible with the original jQuery plugin?  Or should they also call `balanceText.polyfill()` (explicit) if they want polyfill behavior as well?

Rycochet commented:
@BrianGenisio I think they're totally separate things except for compatibility reasons. The polyfill should almost be automatic until it's supported in all browsers (maybe with a query parameter on the `<script>` tag to disable) - the only reason to not enable the polyfill but still include the script is for performance where the user knows it's not needed etc. Sadly not enough people use modern browsers so the "best" way to do things can't be done yet :-/

bfred-it commented:
Now that I think about it, perhaps we can do away with the jQuery wrapper altogether. People can already do `balanceText.balanceText($elements)`. jQuery adds _no value no more._

On a related note, if we can start from scratch and forget jQuery conventions:

```js
// simple, verb-like, enables auto/polyfill mode
balanceText();

// act once
balanceText(el);
balanceText([el, el]);
balanceText($els);
balanceText('.el');

// act now and on resize; also remembers the element for future refresh
balanceText(el, {watch: true});
balanceText('.el', {watch: true}); // selector is remembered as is

// run balance text on any watched elements/selectors
balanceText.updateWatched();
```

IF we want to make the polyfill somewhat automatic, we look for a `data-balance-text` somewhere and maybe `data-balance-text="custom-selector"` and enable it that way (in addition to the manual `balanceText(/* no args */)`)

Rycochet commented:
There's one single value to jQuery - method chaining - the jQuery code would effectively be as simple as:

```javascript
balanceText.addTo = function(obj) {
   obj["balanceText"] = function() {
      var args = [].slice.call(arguments);

      args.unshift(this);
      balanceText.apply(undefined, args);
      return this;
   };
};

balanceText.addTo(jQuery.fn);
```

Making the wrapper into an `.addTo()` call means it can be added to other things and makes code neater and easier to trace and use.

Apart from that my only suggestion is to have `{watch:true}` as the default, add an expando value to any balanceText watched elements, then use a TreeWalker to update (almost as fast as a querySelectorAll, and safer than trying to remember which elements are watched).

BrianGenisio commented:
If it were my project, I'd eliminate all references to jQuery in this repo and create a new repo called `jquery-balance-text` to mirror `angular-balance-text` and `react-balance-text` and the jQuery package can do whatever it needs to maintain compatibility.  I would not assume the polyfill by default, but have an option to polyfill.

It's not my project, however.  I wanted to state my opinion first, and further say that I want to be a good citizen with this PR and make everyone (especially @redmunds) happy.

@bfred-it Yeah, I didn't spend any time making the API look good.  I love your suggestion for how the API should look.  I'll morph the interface to be kinder.  To my point above, I would probably not support `balanceText($els)` since `balanceText([el, el])` works.  The less knowledge about jQuery, the better IMO.

bfred-it commented:
I agree with all your points.

As for supporting `$els` it's actually easy. Call `nodeListAsArray` if `elements` is array-like (rather than `if it's a nodeList`). Done.

Recently for my modules I forwent `.forEach` entirely and just reverted to [plain old `for` loops.](https://github.com/bfred-it/object-fit-images/blob/master/index.js#L180-L193)

Rycochet commented:
@bfred-it NodeList-alike needs more checks sadly - strings, Functions and Window (IE8 at least) have a length, the code I put in Velocity is here - https://github.com/julianshapiro/velocity/blob/master/velocity.js#L556-L565 - note that it's not checking for Window, but it should be relatively understandable.

```javascript
/* Determine if variable is an array-like wrapped jQuery, Zepto or similar element, or even a NodeList etc. */
/* NOTE: HTMLFormElements also have a length. */
isWrapped: function(variable) {
   return variable
      && Type.isNumber(variable.length)
      && !Type.isString(variable)
      && !Type.isFunction(variable)
      && !Type.isNode(variable)
      && (variable.length === 0 || Type.isNode(variable[0]));
};
```

`.forEach` is more readable, but sadly significantly slower - https://jsperf.com/for-vs-foreach/9

bfred-it commented:
@Rycochet it depends on what input you expect. The whole thing already breaks if you pass functions and numbers to it, so that was besides the point. In my `for` example I do check for strings. The only additional check here would be `isNode`

Rycochet commented:
@bfred-it It's a case of catching mistakes and not trusting users - when they accidentally do `balanceText(getNodes);` instead of `balanceText(getNodes());`, or they want to do it on a `<form>` element directly. It's why I've moved over to TypeScript for all professional work so I (hopefully) don't make that sort of error ;-)

BrianGenisio commented:
I just pushed a change to the interface, following the one that @bfred-it suggested.  SO MUCH NICER.  Thank you.

*It should be noted* that this is not a polyfill in a strict sense (it never has been).  It does not look for elements with the style of `text-wrap: balanced`.  It looks for something with a *class name* of `.balance-text`.  I never fully understood that, so I'm sharing with the rest of you, in case that changes any opinions.

I haven't touched the way jQuery is being integrated (currently attaches the legacy interface automatically if window.jQuery exists) when @redmunds has a chance to weigh in.  To summarize, our options are:
1. Automatically attach as a jQuery plugin but require the user to call `$.balanceText()` if they want to "polyfill"
2. Create a `balanceText.addTo(jQuery.fn)` to hook the plugin manually
3. Create two different dist files (via Webpack).  The first is a Vanilla, the second includes Vanilla and wraps it in a legacy jQuery plugin
4. Remove all jQuery compatibility from this repo.  Fork this repo as `jquery-balance-text` and update it to rely on this Vanilla project

redmunds commented:
@BrianGenisio @bfred-it @Rycochet Thanks for all of the input here!

I originally intended this to be a polyfill, but (as noted) never got around to finding/writing code that searches styles applied to document. At this point, I'm happy with whatever is most useful to the community.

Getting rid of jQuery altogether would be awesome. I just didn't want the migration to be too painful. But, there's really no new functionality here, so people can stick with previous version if they don't want to update their code.

Does that map closest to option (2) ?

Let me know if I missed any other questions :) Also, let me know if it would be helpful to pull this in to a (non-master) branch.

bfred-it commented:
Yep, 2 sounds good! (or nothing at all)

non-master sounds good, I might send some PRs based on that releasing v3.0

BrianGenisio commented:
Getting rid of jQuery all together would be closer to #4, I think.  Get rid of any reference to jQuery in 3.0.0 but instead of forking and maintaining a `jquery-balance-text` plugin, we just put something at the top of the README.md that says "If you are looking for the jQuery plugin, please use Version 2.0.0" and tag the repo in it's current place as `2.0.0-jquery` or something.

If we went with option #2, we'd be maintaining two separate interfaces to the same library.  While that may be good for legacy usage, you made a good point: there are no functional differences.  It does keep jQuery users from any future enhancements, but I think the migration from jQuery to this new interface is really straight-forward.

Unless I get any objections, I'm just going with option #5: "goodbye jQuery"

redmunds commented:
Maybe just add a sample of "How to use with jQuery".

@BrianGenisio Let me know when you've pushed all of your current changes and I'll pull your remote branch into a local branch.

BrianGenisio commented:
@redmunds @bfred-it Done.  I updated the README.md to include a little guide for migrating from jQuery (showing equivalent interface calls).  I searched through the entire repo and found a few more references to jQuery.  Now, the ONLY reference to jQuery is in the README.md file in one section.

I have a CDN reference to 3.0.0 AND 2.0.0 for legacy needs.  The path of the CDN files changed a bit, but haven't been published yet, so @bfred-it when you get to the publishing step, you might need to update the CDN links.

Otherwise, I think all open items are complete.  It is ready for @bfred-it to publish, unless anyone has anything else (and I'll jump on it quickly)

Thanks all.  This has been a VERY pleasant interaction :)  I'm excited to get it into production :)

redmunds commented:
@BrianGenisio @bfred-it I pushed this branch to local branch `Khan/native`.

redmunds commented:
Auto-merged via #97
